### PR TITLE
Guided onboarding: add intent based plans

### PR DESCRIPTION
**Pull request by alshakero**
Related to 7019-gh-Automattic/dotcom-forge

## Proposed Changes

* Add mapping to show relevant plans based on user segment.

It should follow this table

image

P2 Segment to Plan mapping: pdDR7T-1xi-p2

## Why are these changes being made?
See pdDR7T-1xi-p2

## Testing Instructions

1. Go to /start/guided
2. The answers to the survey should follow the following tables and the plans should match

Intent (Q1) | Goals (Q2) | Segment | Plans or action
-- | -- | -- | --
Migrate or import | n/a | Migration | Redirect to import
Create for clients | n/a | Developer / Agency | Creator, Entrepreneur, Enterprise
Create for self | Unknown (question skipped) | Unknown | All plans
Create for self | includes Get a website built quickly | DIFM | Redircet to DIFM
Create for self | Sell something and not Get a website built quickly | Merchant | Creator, Entrepreneur, Enterprise
Create for self | not Get a website built quickly and not Sell something | Consumer / Business | Starter, Explorer, Creator
Create for self | Publish a blog | Blogger | Free, Starter, Explorer, Creator
Create for self | Educational/nonprofit | Nonprofit | Free, Starter, Explorer, Creator

### CODE REVIEW

File: packages/plans-grid-next/src/hooks/data-store/types.ts
Comment (chriskmnds) [inline packages/plans-grid-next/src/hooks/data-store/types.ts]: Let's clarify what this property is and why it needs to be introduced in the grid. Also the signature is too open ended. Have you considered defining a more concrete type for the values?

**Review by chriskmnds — CHANGES_REQUESTED** [review]
There is context being introduced into the grid, which i'm not certain it needs to live in the package. Left a comment in the code to get started, but need to take a closer.

File: packages/plans-grid-next/src/hooks/data-store/use-segment-based-plans.ts
Comment (chriskmnds) [inline packages/plans-grid-next/src/hooks/data-store/use-segment-based-plans.ts:1]: This hook needs to go/live outside of `plans-grid-next`. Part of our goal with this package is to keep it as lean as possible, with context like "answers to a survey" not really being part of anything. Also, any new properties, whether on the grid components or its helpers/hooks to be of generic nature, not context specific (e.g. `segmentationAnswers` is of the latter).

I think there are a couple of options:

1. Define a `useIntentFromSegmentationAnswers` and return a different `PlansIntent` for every combination of plan types - so you'd be returning `plans-guided-default`, `plans-guided-blogger`, `plans-guided-consumer`, etc. Hopefully, this also serves as a signal to define a better `Segment` type for the survey work - cause it's all strings right now. You'd introduce those intents in the main switch block of `use-grid-plans` to define the plan type combinations (as done for everything else - no other logic/change there). The hook would live either near the logic or somewhere else in Calypso e.g. in `plans-features-main`.
2. Define a `usePlanTypesWithIntentOverride( { intent } )` (the name kinda slips to me now) and pass that into the hooks. In there, you can call this hook and return whatever type combination for the "one" intent defined. We'd somehow work that into the grid hooks as an optional override.

I'll give it some more thought/perspective, but these are options, even if intermediate.

Comment (chriskmnds) [reply packages/plans-grid-next/src/hooks/data-store/use-segment-based-plans.ts:1]: p.s. the difference between multiple intents and single intent doesn't seem to be significant right now. So either option doesn't seem to deviate too far. If in a later scenario, you consider "flow divergence" (so diverging to a different flow on every answer) - which would give you a different "site intent" in the end (I think!) - then the first option is probably better due to mapping across to "plan intent" already.

File: packages/plans-grid-next/src/hooks/data-store/use-segment-based-plans.ts
Comment (chriskmnds) [inline packages/plans-grid-next/src/hooks/data-store/use-segment-based-plans.ts]: I guess still WIP? This doesn't exist in the package, right❓ (I hope not, we'd need to refactor out now 😬 )

File: packages/plans-grid-next/src/hooks/data-store/use-segment-based-plans.ts
Comment (chriskmnds) [reply packages/plans-grid-next/src/hooks/data-store/use-segment-based-plans.ts:1]: cc @Automattic/martech-decepticons

File: packages/plans-grid-next/src/hooks/data-store/use-segment-based-plans.ts
Comment (escapemanuele) [reply packages/plans-grid-next/src/hooks/data-store/use-segment-based-plans.ts]: Still WIP yes, the hook now lives in another part https://github.com/Automattic/wp-calypso/pull/91017/files. 

@agrullon95, @chriskmnds is right, as our hook to retrieve the segment is not correlated to the plans, but to be used in various places, so we need to find another place for it :)

Comment (escapemanuele) [reply packages/plans-grid-next/src/hooks/data-store/use-segment-based-plans.ts]: Maybe `packages/onboarding`?

Comment (chriskmnds) [reply packages/plans-grid-next/src/hooks/data-store/use-segment-based-plans.ts]: Maybe that's fine? 

We won't be querying/using that package in `plans-grid-next` either, so these still need to be pulled in differently here. The package is imported here, but I think it should be safe to remove - it's only for some styles that we may have cleaned up already.

File: packages/plans-grid-next/src/hooks/data-store/use-segment-based-plans.ts
Comment (southp) [reply packages/plans-grid-next/src/hooks/data-store/use-segment-based-plans.ts:1]: What @chriskmnds suggested above should already work pretty well. Fwiw I'd like to throw in another option to consider :) 

The general idea of the classic signup framework is that it's a pipeline filling up the required "depedency data" through calls to the [submitSignupStep](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/signup/progress/actions.js#L127) function, e.g. [how the domains step does it](https://github.com/Automattic/wp-calypso/blob/trunk/client/signup/steps/domains/index.jsx#L134). That also means a former step can produce data that can be consumed by a latter step to act on accordingly. e.g. again see [how the domain step gets the design type data](https://github.com/Automattic/wp-calypso/blob/trunk/client/signup/steps/domains/index.jsx#L585) set by a former step so it can pick different TLDs accordingly.

So, if the segmentation survey was a new signup step, one more way to do it is to:

1. Let it submit a new dependency data entry, e.g. `segmentationSurveyAnswers`. 
2. In the plans step component, derive the intent in some ways using `segmentaionSurveyAnswers` in the `plansFeaturesList` function before feeding it into PlansFeaturesMain: https://github.com/Automattic/wp-calypso/blob/trunk/client/signup/steps/plans/index.jsx#L144. 

A slight benefit of doing this way is that the step submission will be tracked by default using `calypso_signup_step_submit` event, so you won't need to setup additional tracking.

I don't have strong preference on either way; just want to raise another option here in case it helps :)

Comment (chriskmnds) [reply packages/plans-grid-next/src/hooks/data-store/use-segment-based-plans.ts:1]: @southp yep. it sounds like a good way to go (keep it near the step logic and pass it into `PlansFeaturesMain` via the prop)

> The hook would live either **near the logic** or somewhere else in Calypso e.g. in plans-features-main.

I think what you outlined would precisely cover the "near the logic" part 😁

File: packages/plans-grid-next/src/hooks/data-store/use-grid-plans.tsx
Comment (alshakero) [inline packages/plans-grid-next/src/hooks/data-store/use-grid-plans.tsx]: @chriskmnds this is the only pollution to the package now.

File: packages/plans-grid-next/src/hooks/data-store/use-segment-based-plans.ts
Comment (alshakero) [reply packages/plans-grid-next/src/hooks/data-store/use-segment-based-plans.ts:1]: That's a great idea. I would love to submit the survey answers via the `submitSignupStep` function in a next PR. I don't love the fact that we have to post and fetch between step 1 and 3.

File: client/my-sites/plans/hooks/use-segmented-intent.tsx
Comment (agrullon95) [inline client/my-sites/plans/hooks/use-segmented-intent.tsx]: We should return this format, so we can use the fetch state for the plans page. This would prevent the plans page from flashing the previous plans before updating.
```return { segment: fallback, isFetchingSegment: isFetching };```

**Review by agrullon95 — APPROVED** [review]
The segment -> plans mapping is working correctly. One minor change for this PR or follow-up PR is taking into account the loading state for the plans page until the segment is fetched.

I left an example as a comment in the code. WDTY?

File: client/my-sites/plans/hooks/use-segmented-intent.tsx
Comment (alshakero) [reply client/my-sites/plans/hooks/use-segmented-intent.tsx]: Oh yeah, that's a great idea. Would you mind pushing that?

Comment (agrullon95) [reply client/my-sites/plans/hooks/use-segmented-intent.tsx]: I pushed the segment loading state changes.

File: client/my-sites/plans/hooks/use-segmented-intent.tsx
Comment (alshakero) [inline client/my-sites/plans/hooks/use-segmented-intent.tsx]: We should add another `if` before this, that handles `! enabled` and return `isFetchingSegment: false`.

Comment (agrullon95) [reply client/my-sites/plans/hooks/use-segmented-intent.tsx]: Updated

File: client/my-sites/plans/hooks/use-segmented-intent.tsx
Comment (chriskmnds) [inline client/my-sites/plans/hooks/use-segmented-intent.tsx]: (Nit/Non-blocking) Why `blogId` and not `siteId`? Also, why the default `0`?

Ideally, I'd expect any consumer of `siteId`  to just respect the original semantics and reason organically with `undefined | null` values. The selected-site ID (so `siteId`) has a fairly standard meaning and form - we don't need dummy concrete defaults like `0` or `-1` whenever we pull it into our logic/code.

File: client/my-sites/plans/hooks/use-segmented-intent.tsx
Comment (chriskmnds) [inline client/my-sites/plans/hooks/use-segmented-intent.tsx]: Is it necessary to have this `enabled` flag? Can you not return `{ segment: undefined, isFetchingSegment: false }` when no segmentation survey answer exists?

File: packages/plans-grid-next/src/types.ts
Comment (chriskmnds) [inline packages/plans-grid-next/src/types.ts]: Is this actually effective for anything? If not, then I think best to remove it. Ideally, the intents are values to tailor the plan types used in `useGridPlans`, and shouldn't exist just for external context/purpose (e.g. to control logic outside of plans-grid-next). It's likely the source of other complexities I'm seeing in the code (so also the basis of my other comments I think).

File: packages/plans-grid-next/src/types.ts
Comment (chriskmnds) [inline packages/plans-grid-next/src/types.ts]: WDYT about `plans-guided-segment-xyz`, so being a little more specific with these?

File: client/my-sites/plans-features-main/index.tsx
Comment (chriskmnds) [inline client/my-sites/plans-features-main/index.tsx]: This will likely not be necessary after removing `plans-guided`. In any case, I would rethink returning early here. If this lingers, let me have a moment totake a closer look before merging pls.

File: client/my-sites/plans/hooks/use-segmented-intent.tsx
Comment (chriskmnds) [inline client/my-sites/plans/hooks/use-segmented-intent.tsx]: It doesn't sound like we will need this `fallback` logic. Hopefully, removing `plans-guided` will render this unnecessary. 

It also introduces a bit of indirection with the default value here when undefined previously... it kinda breaks the data flow a bit.

**Review by chriskmnds — DISMISSED** [review]
Thanks for addressing the feedback. I left a few more comments/concerns if you can take a look pls. :-)

File: client/my-sites/plans/hooks/use-segmented-intent.tsx
Comment (alshakero) [reply client/my-sites/plans/hooks/use-segmented-intent.tsx]: The reason is that the survey API calls it `blog_id` and since this hook is an extension of the API, I kept it consistent with the API.

The reason for the zero is also the API. You can see the response below. The survery endpoint assigns `0` when the user has no selected site.

```js
{
    "body": [
        {
            "blog_id": 0, // <----- this
            "answers": {
                "what-are-your-goals": [
                    "blog"
                ],
                "what-brings-you-to-wordpress": [
                    "myself-business-or-friend"
                ]
            }
        }
    ],
    "status": 200,
    "headers": {
        "Allow": "GET"
    }
}
```

I'll make the function handle falsies differently.

File: client/my-sites/plans/hooks/use-segmented-intent.tsx
Comment (alshakero) [reply client/my-sites/plans/hooks/use-segmented-intent.tsx]: Yes. The enabled also saves us unneeded `fetch` requests when the query is disabled.

File: packages/plans-grid-next/src/types.ts
Comment (alshakero) [reply packages/plans-grid-next/src/types.ts]: Man, I love your wide angle when viewing things. Great feedback. Removed.

File: client/my-sites/plans/hooks/use-segmented-intent.tsx
Comment (alshakero) [reply client/my-sites/plans/hooks/use-segmented-intent.tsx]: > It doesn't sound like we will need this fallback logic. 

~~So the issue is that this hook only filters the plans grid _in some cases_. I'm not sure how to opt-out of the filtering without a fallback value.~~ Removed the mechanism. You're right.

File: client/my-sites/plans-features-main/index.tsx
Comment (alshakero) [reply client/my-sites/plans-features-main/index.tsx]: I updated it to

```
if ( flowName === 'guided' ) {
	return setIntent( intentFromSegmentationSurvey );
}
```

WDYT? I'm not sure about removing the early return. Doesn't seem safe to engage other logic that may change the intent.

Comment (alshakero) [reply client/my-sites/plans-features-main/index.tsx]: I guess you are right. I removed the `if` entirely, since the hook now just returns `undefined` when it's not activated.

Comment (chriskmnds) [reply client/my-sites/plans-features-main/index.tsx]: Hey Omar. I can review the latest tomorrow and give you more concrete feedback. 

For this hook, it's a fairly generic piece of logic that controls the whole intent system in Calypso e.g. encodes a strong priority: `intent from props` > `site intent` > `default`, with the only exception being that a user may click somewhere to show the default grid, hence force the default intent. That functionality should be preserved as a bit of a golden rule. We've had fairly sporadic requests to "show all plans" and I don't doubt more will surface (including for the "guided" flow work). I would not start adding complexity and early returns here.

I think what you are looking for is to pass an intent into `plans-features-main`, so maybe move some of the logic a layer above. What this call is doing is not that different from what we mention in the signature to avoid doing:

```
intent: intentFromProps, // do not set a default value for this prop here
```

I will give it a closer look tomorrow.

Comment (alshakero) [reply client/my-sites/plans-features-main/index.tsx]: Hey. Thanks for elaborating! I already pushed a revised version before seeing your comment. Would love to hear your thoughts, I think you're more likely to agree with the current version.

**Review by escapemanuele — APPROVED** [review]
Tested all the scenarios and the plans are showing accordingly 👍

File: client/my-sites/plans-features-main/index.tsx
Comment (chriskmnds) [inline client/my-sites/plans-features-main/index.tsx]: I may be missing something but is there anything blocking from moving this call to the consumer of `plans-features-main` and avoid the unnecessary logic/conditioning in the main hook below? Just pass the intent you want into `plans-features-main`. That's what the `intent` prop is meant for.

You can still call it here to get the `isLoadiingSegment` state if that's meaningful to anything, but I'd rather we avoid `intentFromSegmentationSurvey` being defined explicitly here. You are overriding the `intentFromProps` indirectly with this. I would not expect any changes to the `useEffect` hook below 🤷‍♂️

File: client/my-sites/plans-features-main/index.tsx
Comment (chriskmnds) [reply client/my-sites/plans-features-main/index.tsx]: Thanks. Replied [here](https://github.com/Automattic/wp-calypso/pull/90809/files#r1620213647)

**Comment by heavyweight** [comment]
Love how much simpler this got

File: client/my-sites/plans-features-main/index.tsx
Comment (alshakero) [reply client/my-sites/plans-features-main/index.tsx]: That's when I started, really. But the parent component is a class component, and doing network requests there sucks. I'd have to make Redux actions, selector, reducer, etc... I can do all of these, but IMHO the trade-off is not worth it. 4 lines vs a bunch of files.

> You are overriding the intentFromProps indirectly with this. 

That's a great point! I wasn't worried about it because the hook only engages for this flow. But it's still code smell to ignore props.

How about [this](https://github.com/Automattic/wp-calypso/pull/90809/commits/d9636d4d6eae209764884de091e69114c4bde322)?

File: client/my-sites/plans-features-main/index.tsx
Comment (southp) [inline client/my-sites/plans-features-main/index.tsx]: It looks like we are deriving the intent value according to the segmentation survey here. There are two main drawbacks here:

1. Once a site is created, the intent value will be stored in the stie meta. e.g. that's how a site created from `/setup/newsletter` will see the same plan mix at the logged-in `/plans`. Ideally all the survey data handling should only happen in the signup flow. After that, the site should just use the stored meta data.
2. An extra `GET` request to the survey endpoint will be fired everytime `<PlansFeaturesMain>` is rendered no matter it's needed or not, e.g. an existing account without ever seeing the survey visits `/plans`. 

I have a feeling that it might be something that can be addressed together with @chriskmnds 's [comment below](https://github.com/Automattic/wp-calypso/pull/90809/files#r1620213647) since it's likely also about pulling this to the consumer end. If not, it'd be wonderful if we can address as a follow-up :)

Comment (alshakero) [reply client/my-sites/plans-features-main/index.tsx]: Finally I see the light. Thank you so much @southp and @chriskmnds! Pushed.

**Comment by alshakero** [comment]
**Note:** don't merge because this introduces a regression. The survey answers are now carried over from the guided flow into the following visits of the plans grid in other flows.

File: client/my-sites/plans-features-main/index.tsx
Comment (chriskmnds) [reply client/my-sites/plans-features-main/index.tsx]: Hey Omar. I haven't had a chance to check the last update. Concerning the class component though, you can write a simple HOC in that case, call the hook and propagate the value in the props. It can all be done in the same file, 3-4 lines wrapping the connected component, no?

Comment (chriskmnds) [reply client/my-sites/plans-features-main/index.tsx]: ^ that's been our strategy for cases in plans-grid that we needed the interplay of hooks and classes, before refactoring everything to function form

I think here you can find examples of a wrapper component or hoc to follow: https://github.com/Automattic/wp-calypso/blob/0d716ccd7550e3a7a4abd6bc6d3b0ad1b2db4540/client/my-sites/plan-features-2023-grid/index.tsx#L1060

File: client/signup/steps/plans/index.jsx
Comment (southp) [inline client/signup/steps/plans/index.jsx:122]:
```suggestion

// TBD:
// The reason that we need to narrow down the effect here by checking the flow name is because the signup progress is shared among flows. However, that behavior is problematic here
// It used to cause another issue here: https://github.com/Automattic/wp-calypso/pull/74329
// We need to find a better way to define a signup session with cleaner boundary of shared storage.
```

Great catch. It's a framework-limitation that will need some more careful thoughts to fix, so this looks like a good workaround for now. I've also suggested to add an inlined TBD note for posterity.

Comment (escapemanuele) [reply client/signup/steps/plans/index.jsx:122]: Thank you @southp. We will test this PR for regressions and then probably merge on Monday

Comment (southp) [reply client/signup/steps/plans/index.jsx:122]: Awesome, @escapemanuele . To better track this, I've created a tracking issue for it: https://github.com/Automattic/wp-calypso/issues/91377. Thanks to the work here, we now have one solid instance to check when we propose a solution to it :)

Comment (chriskmnds) [reply client/signup/steps/plans/index.jsx:122]: I think the issue here stems from the solution/approach, not the framework's limitation (which does exist, but it's kinda concequential that it surfaces). 

If plans-intent was a property on the flow, not something computed dynamically, then we'd propably not need to bother at all. So at a higher level, instead of deriving intent dynamically, send user to a different flow that has that intent configured. I proposed this earlier as "flow divergence", and seeing that it's only relevant for a couple of intents, then it sounds more tempting now, unless any other underlying concerns. 

‐--------

p.s. https://github.com/Automattic/wp-calypso/issues/91377 brings a bit of a "fix the symptom" vibe to me, but will comment there 🤔

Comment (alshakero) [reply client/signup/steps/plans/index.jsx:122]: I'm worried about the complexity that will emerge in tracks and the A/B/C experiment we'll run if we split this into 5 flows. I'll merge this, but please don't take it as the end of accepting (and appreciating) feedback!

**Pull request merged**
